### PR TITLE
feat: add implementation rule

### LINKED ISSUE #4: Validate presence of implementation on camunda:ServiceTaskLike
#### The rule should detect the following modeling patterns

* A `camunda:ServiceTaskLike` (service task, send task, message throw event) is missing implementation

#### How does the rule improve the BPMN diagram?

The engine will throw an error if implementation details are not provided. Showing the message early gives users the chance to prepare their diagrams and saves them a few roundtrips against the engine.

#### Rule Details

* Name: `implementation`
* Default notification level: `warning`

#### What alternatives did you consider? 

Always doing direct validation against the engine would obsolete this rule.

### PR DESCRIPTION
**Pull request by barmac**
Closes #4

### CODE REVIEW

**Comment by barmac** [comment]
~~On the other hand, the implementation details are not rendered on a diagram, so it doesn't seem to make sense to use pictures to visualise the rule. Still, there is no option in the tool to opt-out from generating images for selected diagrams.~~

The documentation does not include an image due to the fact that difference is not rendered on the diagram. This is also the reason why I resigned from the diagram examples and used code snippets instead.

File: rules/implementation.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ * Rule that reports the usage of collapsed sub-processes.
+ */
+module.exports = function() {
+
+  function check(node, reporter) {
+    if (is(node, 'camunda:ServiceTaskLike')) {
+
+      // quirk: bpmn:MessageEventDefinition is a camunda:ServiceTaskLike
+      // even though it's not a bpmn:Task
+      if (is(node, 'bpmn:MessageEventDefinition')) {
```
Comment (nikku) [inline rules/implementation.js]: Why is this one excluded? It does require an implementation to work (if embedded in a throwing message event). Or is it optional from the engine perspective?

File: rules/implementation.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        hasAnyAttribute(node, implementationAttributes)
+      ) {
+        return;
+      }
+
+      if (is(node, 'bpmn:BusinessRuleTask') && hasAttribute(node, 'camunda:decisionRef')) {
+        return;
+      }
+
+      reporter.report(node.id, 'Implementation should be specified');
```
Comment (nikku) [inline rules/implementation.js]: I'd rather use `XY is missing here` to better align with our existing wording.

**Review by nikku — CHANGES_REQUESTED** [review]
Two small comments.

File: rules/implementation.js (hunk 1 above)
Comment (barmac) [reply rules/implementation.js]: I've tested this with the engine and it accepts both intermediate send event as well as message end tasks without implementation.

Comment (barmac) [reply rules/implementation.js]: Maybe it's a bug on the engine: https://docs.camunda.org/manual/7.15/reference/bpmn20/events/message-events/#camunda-extensions-for-messageeventdefinition

Comment (barmac) [reply rules/implementation.js]: Let's be more strict; I'll remove this condition.

Comment (nikku) [reply rules/implementation.js]: Could you double check with the Engine folks first? Let's support what is implemented, not what is supposed to be fixed.

If the engine supports "none" message throw events, that is fine for me, too.

Comment (nikku) [reply rules/implementation.js]: I'm aware of the fact that [this _is_ the way to make these throw events executable](https://github.com/camunda/bpmnlint-plugin-camunda/pull/11#discussion_r615736426). But if no implementation does not result in any error, that should be fine for us, too.

Comment (barmac) [reply rules/implementation.js]: Yeah well. My tests revealed that the events were just treated as none events. I'll re-implement the type check and let the engine people know about the _mandatory but not mandatory_ behaviour of the engine.

Comment (barmac) [reply rules/implementation.js]: I've started the dicussion [here](https://camunda.slack.com/archives/C6M6X3EKZ/p1618831753091200). I think we can change the rule when/if the engine starts to reject deployments with missing implementation. Until then, we can keep the less strict version.

File: rules/implementation.js (hunk 2 above)
Comment (barmac) [reply rules/implementation.js]: `Implementation is missing`

File: package.json [hunk 3]
```diff
@@ -28,7 +28,9 @@
   "license": "MIT",
   "devDependencies": {
     "@bpmn-io/bpmnlint-generate-docs-images": "^0.2.0",
+    "bpmn-moddle": "^7.0.4",
     "bpmnlint": "^7.2.1",
+    "camunda-bpmn-moddle": "^5.1.2",
```
Comment (nikku) [inline package.json:33]: What is generally interesting about this one is the fact that the new rule will ever only work if the `moddle` has been loaded with it. I.e. on the CLI, which right now does not have a way to specify the moddle packages being used, `implementation` won't ever be properly validated.

Is that an issue? Not sure.

Comment (barmac) [reply package.json:33]: That's true. `bpmnlint` as a CLI should somehow pick up the additional schemas, but it's not implemented yet.

Still, this is probably outside of the scope of this issue. The rule will work correctly in any environment where `camunda-bpmn-moddle` is used and this is the only environment where the rule is needed.

I'll have an issue for the `bpmnlint` repo.

Comment (barmac) [reply package.json:33]: https://github.com/bpmn-io/bpmnlint/issues/62

**Comment by barmac** [comment]
Got feedback from @ThorbenLindhauer that we should disallow missing implementation for the Message Throw/End Events. So this would be a way to promote best practices.

From readme:

> Camunda modeling guidelines, packed as a bpmnlint plug-in.

In that case, I think it makes sense to be more strict than the engine.

File: package.json (hunk 3 above)
Comment (nikku) [reply package.json:33]: Thanks for opening that issue. I'm fine to proceed with this PR. We can improve via https://github.com/bpmn-io/bpmnlint/issues/62 as we find time.

**Comment by nikku** [comment]
Let's implement the strict message event definition validation (https://github.com/camunda/bpmnlint-plugin-camunda/pull/11#issuecomment-823113289) and we're good I think!

**Comment by barmac** [comment]
Some insights about why they changed the outline: https://blog.chromium.org/2020/03/updates-to-form-controls-and-focus.html#Blog1:~:text=A%20More%20Visible%20Focus%20Ring

File: package.json [hunk 4]
```diff
@@ -7,7 +7,7 @@
     "all": "run-s lint test",
     "lint": "eslint .",
     "dev": "npm test -- --watch",
-    "test": "mocha 'test/**/*.js'",
+    "test": "mocha test/rules/**/*.js test/**/*-spec.js",
```
Comment (nikku) [inline package.json]: What exactly is this for?

**Review by nikku — COMMENTED** [review]
I think you removed most of the rough edges. 

One thing I thought about: How much sense does it make to activate this rule in non-executable processes? 

In BPMN we distinguish between these. Non executable processes are meant for documentation only and not picked up by the engine. I'd assume I can do anything in there, with no such limit.

**Comment by barmac** [comment]
Yeah, that makes sense. I haven't thought about it. I'll extend the test case to not bug the user about non-executable processes.

File: package.json (hunk 4 above)
Comment (barmac) [reply package.json]: I wanted to exclude the utils from being run. But maybe it's a better idea to always use `-spec` suffix.

Comment (barmac) [reply package.json]: Reorganized via https://github.com/camunda/bpmnlint-plugin-camunda/pull/11/commits/5c5da9871b1f30917f4858f4add1fca01b3d6e2d

**Comment by barmac** [comment]
Now we ignore non-executable processes.

**Review by nikku — APPROVED** [review]
Very nice :cake: :tada:

**Pull request merged**